### PR TITLE
Add $$ASSIGN directive so we can change values inside a $$LOOP. 

### PR DESCRIPTION
**Pull request by mcb64**
This will help template the mpod IOC, in particular, which needs a total number of channels across all boards.

### CODE REVIEW

File: expand.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
 inc     = re.compile("^\$\$INCLUDE\((.*)\)")
 idxre   = re.compile("^INDEX([0-9]*)")
 doubledollar = re.compile("^(.*?)\$\$")
-keyword      = re.compile("^(ROOT|SUBSTR|UP|LOOP|IF|INCLUDE|TRANSLATE|COUNT|NAME)\(|^(CALC|IFCALC)\{")
+keyword      = re.compile("^(ROOT|SUBSTR|UP|LOOP|IF|INCLUDE|TRANSLATE|COUNT|NAME)\(|^(ASSIGN|CALC|IFCALC)\{")
```
Comment (ZLLentz) [inline expand.py:33]: Unrelated to my review: maybe a future version of the IOC templating could leverage standard tools like jinja instead of using custom $$ macros. Jinja has essentially any option we'd ever want when it comes to assigning variables, doing loops, if statements, calculations, etc.: https://jinja.palletsprojects.com/en/3.1.x/templates/

I'll review the PR in a bit without thinking too much about the above.

Comment (aberges-SLAC) [reply expand.py:33]: On that tangent @ZLLentz, are there any ongoing efforts to upgrade `expand.py2` to `expand.py3`? I have a local copy that I was working with to better comprehend the script and shoved it through `2to3` as a first pass. I have some ideas for simplifying some of the regex for parsing purposes and want to test it more in my local dev area before I claim any sort of improvement (_I anticipate I'm breaking at least one thing_)

Comment (ZLLentz) [reply expand.py:33]: There haven't been any ongoing efforts to update this repo to py3. There isn't a huge amount of short-term benefit for such an effort unless we're pairing it with some other capability expansions.

The default python os install on rhel7 is python2, so when you `make` we have some guarantee that the interpreter is available.

Perhaps the story changes when we move to rocky9, which has only python3 installed, and perhaps now with the conda setups we could update it without too much hassle. In fact, with rocky9, it may become important to update to py3...

File: expand.py [hunk 2]
```diff
@@ -678,6 +688,19 @@ def expand(cfg, lines, f, isfirst=False):
                     except:
                         cnt = "0"
                     f.write(cnt)
+                elif kw == "ASSIGN":
```
Comment (ZLLentz) [inline expand.py:735]: Despite the diff being concise, this code is hard to follow and review in general because the script doesn't have any docstrings and only some sections have comments. Maybe the most important thing missing here is some sort of documentation: how would one go about using the new ASSIGN directive? Something like `$$ASSIGN{VARNAME,4}`?

File: expand.py [hunk 3]
```diff
@@ -678,6 +688,19 @@ def expand(cfg, lines, f, isfirst=False):
...
+                    args = argm.group(1).split(",")
+                    output = StringIO.StringIO()
+                    expand(cfg, [args[1]], output, isfirst)
+                    value = output.getvalue()
+                    output.close()
+                    try:
+                        v = cfg.eval_expr(value)
```
Comment (ZLLentz) [inline expand.py]: Does this mean we're allowing arbitrary calculation expressions in the ASSIGN directive? If so, is cfg giving them access to some sort of shared namespace of values to pull from? This would be useful to document.

File: expand.py [hunk 4]
```diff
@@ -678,6 +688,19 @@ def expand(cfg, lines, f, isfirst=False):
...
+                    except:
+                        v = 0
```
Comment (ZLLentz) [inline expand.py]: Is this a useful default in some cases? If so, great, if not, then maybe this really should raise an exception and let the user know that they misused the ASSIGN directive.

File: expand.py [hunk 5]
```diff
@@ -576,6 +581,11 @@ def expand(cfg, lines, f, isfirst=False):
                         cfg.ddict = rename_index(olddict.copy())
                         cfg.ddict.update(inst)
                         expand(cfg, t[0], f, isfirst)
+                        # Now, within the $$LOOP, we might have done some $$ASSIGNs.
+                        # We need to pull these back into olddict!
+                        for dname in cfg.assigns:
```
Comment (ZLLentz) [inline expand.py]: Is there any guarantee that cfg.assigns only holds assignments from this loop? I think this could also hold assigns from prior to the first loop during the first evaluation of a loop block, though I guess this probably doesn't cause any issues

**Review by ZLLentz — COMMENTED** [review]
This is a good addition to the macro expander
I asked a few questions to help with my understanding and to request some documentation about how to use the new directive

File: expand.py (hunk 1 above)
Comment (mcb64) [reply expand.py:33]: > Unrelated to my review: maybe a future version of the IOC templating could leverage standard tools like jinja instead of using custom $$ macros. Jinja has essentially any option we'd ever want when it comes to assigning variables, doing loops, if statements, calculations, etc.: https://jinja.palletsprojects.com/en/3.1.x/templates/

Yeah, reinventing the wheel 'r' us... :-(

File: expand.py (hunk 2 above)
Comment (mcb64) [reply expand.py:735]: Exactly.  But actually, the power of this is $$ASSIGN{VARNAME,$$VARNAME+4}, so you can accumulate running totals of things.

This code completely sucks.  As way of explanation, I wrote most of it in an afternoon as a proof-of-concept for templating, but never bothered to fix it up and document what the hell I was doing we actually started *using* this...

File: expand.py (hunk 3 above)
Comment (mcb64) [reply expand.py]: Yes, the cfg is the current name space.  For the most part, this is just the defines in the .cfg file, but it is augmented with additional values when $$LOOPing.  (In particular, $$INDEX is the iteration counter, and if iterating over instantiations, all of the parameters are added to cfg.  We're doing a little dance to move INDEX to INDEX1, INDEX1 to INDEX2, and so on, whenever we start a new loop.)

File: expand.py (hunk 4 above)
Comment (mcb64) [reply expand.py]: It probably isn't useful.  There are a lot of other instances of this type of behavior...

File: expand.py (hunk 5 above)
Comment (mcb64) [reply expand.py]: No, I think you have a point here.  assigns needs to be changed to keep track of the nesting level.  (If  we have two loops, the inner would export to the outer, but the outer would just drop the value on the floor when it went back to the top-level namespace.

Comment (ZLLentz) [reply expand.py]: Right that makes sense, I forgot about nested loops

File: expand.py (hunk 2 above)
Comment (ZLLentz) [reply expand.py:735]: I think in the PR description, in a comment, or wherever this module is documented, including this example would be really helpful.

File: expand.py (hunk 1 above)
Comment (aberges-SLAC) [reply expand.py:33]: > Perhaps the story changes when we move to rocky9, which has only python3 installed, and perhaps now with the conda setups we could update it without too much hassle. In fact, with rocky9, it may become important to update to py3...

Since this is an inevitable future and we only have Mike for a few more weeks, it might be worthwhile to start this sooner vs. later. I can volunteer to make a branch on this and track my changes/efforts/questions until someone more appropriate takes over/has time. Some things like this $$ASSIGN macro and other nuances are very relevant for some changes I want to make to SmarAct IOCs (and maybe integrating into jinja templates, depending on how we want to handle mapping values from additional scripts to set some default PV behavior for controls).

File: expand.py (hunk 2 above)
Comment (mcb64) [reply expand.py:735]: The problem in a nutshell... the mpod uses snmp, which is generally very slow.  So we want to use an snmpwalk to read *all* of the mibs at once.  Now, we know how many mibs per channel there are, and where they are in the address space... but we need to know how many channels we have in total in order to know how many mibs we should read in the walk.  Prior to this, this was impossible... which is largely why the mpod IOC was untemplated.  But now, we can do the following:
```
$$ASSIGN{TOTAL,0}
$$LOOP(BOARD)
...
dbLoadRecords("db/mpod_$$(TYPE)module_$$(COUNT)channel.db", "HOST=$(CRATE),...")
...
$$ASSIGN{TOTAL,$$TOTAL+$$COUNT}
$$ENDLOOP(BOARD)
...
Snmp2cWalk("$(CRATE)", "guru", "WIENER-CRATE-MIB::outputStatus",              $$CALC{11*$$TOTAL}, 1.0)
Snmp2cWalk("$(CRATE)", "guru", "WIENER-CRATE-MIB::outputSupervisionBehavior", $$CALC{10*$$TOTAL}, 1.0)
Snmp2cWalk("$(CRATE)", "guru", "WIENER-CRATE-MIB::outputCurrentRiseRate",     $$CALC{3*$$TOTAL},  1.0)
Snmp2cWalk("$(CRATE)", "guru", "WIENER-CRATE-MIB::outputUserConfig",          $$CALC{3*$$TOTAL},  1.0)
```
Where instances have the form:
```
BOARD(TYPE=lv/hv,COUNT=8/16,...)
```

File: expand.py (hunk 5 above)
Comment (mcb64) [reply expand.py]: In general, we don't care about $$ASSIGNS out of a loop... we'll put them in the first assigns set, but the values go directly into our value dictionary, so there is no need to worry about exporting them anywhere.  The issues all occur when we enter a new context... and the only context we have is entering a $$LOOP.

File: expand.py (hunk 2 above)
Comment (ZLLentz) [reply expand.py:735]: Yeah, this all makes sense and is good and this PR achieves/enables it. My only complaint was essentially "this is a new feature that isn't explained very well and the PR description doesn't describe it and the code isn't self-describing and is there a confluence page for it?"

I'm going to send my review approval either way

**Review by ZLLentz — APPROVED** [review]

File: expand.py (hunk 1 above)
Comment (ZLLentz) [reply expand.py:33]: > it might be worthwhile to start this sooner vs. later

I think you're right about this

File: expand.py (hunk 2 above)
Comment (mcb64) [reply expand.py:735]: Your complaint is totally valid.  If you search for "templated IOC", you will get my page that describes the use of expand.py... and I will add $$ASSIGN here shortly.
